### PR TITLE
MSTEST0032: Always pass assertions 

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/3107

### CODE REVIEW

File: src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+using MSTest.Analyzers.Helpers;
+using MSTest.Analyzers.RoslynAnalyzerHelpers;
+
+namespace MSTest.Analyzers;
+
+/// <summary>
+/// MSTEST0025: <inheritdoc cref="Resources.PreferAssertPassOverAlwaysTrueConditionsAnalyzerTitle"/>.
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs]: Is it `MSTEST0025` rule?

File: src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+public sealed class PreferAssertPassOverAlwaysTrueConditionsAnalyzer : DiagnosticAnalyzer
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs]: Do we have `Assert.Pass`?

File: src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            IArgumentOperation? valueArg = GetValueArgument(operation);
+
+            ITypeSymbol? valueArgType = valueArg?.Value.GetReferencedMemberOrLocalOrParameter().GetReferencedMemberOrLocalOrParameter();
+
+            if (!SymbolEqualityComparer.IncludeNullability.Equals(valueArgType?.OriginalDefinition, nullableSymbol) || valueArgType?.NullableAnnotation != NullableAnnotation.Annotated)
+            {
+                context.ReportDiagnostic(operation.CreateDiagnostic(Rule, operation.TargetMethod.Name));
+            }
+        }
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs]: Same feedback as on the other PR, this should be moved to the `IsAlwaysFalse`.

File: src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                context.RegisterOperationAction(context => AnalyzeOperation(context, assertSymbol, nullableSymbol), OperationKind.Invocation);
+            }
+        });
+    }
+
+    private static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol assertSymbol, INamedTypeSymbol? nullableSymbol)
+    {
+        var operation = (IInvocationOperation)context.Operation;
+        if (assertSymbol.Equals(operation.TargetMethod.ContainingType, SymbolEqualityComparer.Default) &&
+            IsAlwaysFalse(operation, nullableSymbol))
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs]: You want to check `IsAlwaysTrue`, isn't it?

File: src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs (hunk 2 above)
Comment (engyebrahim) [reply src/Analyzers/MSTest.Analyzers/PreferAssertPassOverAlwaysTrueConditionsAnalyzer.cs]: no

File: src/Analyzers/MSTest.Analyzers/Helpers/DiagnosticIds.cs [hunk 5]
```diff
@@ -35,4 +35,5 @@ internal static class DiagnosticIds
     public const string PublicMethodShouldBeTestMethodRuleId = "MSTEST0029";
     public const string TypeContainingTestMethodShouldBeATestClassRuleId = "MSTEST0030";
     public const string DoNotUseSystemDescriptionAttributeRuleId = "MSTEST0031";
+    public const string PreferNotUsingAlwaysTrueAssertConditionsAnalyzerRuleId = "MSTEST0032";
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Helpers/DiagnosticIds.cs]: Maybe something like `ReviewAlwaysTrueAssertConditionAnalyzerRuleId`, please also update the class and resx entries accordingly.

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 6]
```diff
@@ -454,4 +454,10 @@
   <data name="DoNotUseSystemDescriptionAttributeDescription" xml:space="preserve">
     <value>'System.ComponentModel.DescriptionAttribute' has no effect in the context of tests and you likely wanted to use 'Microsoft.VisualStudio.TestTools.UnitTesting.DescriptionAttribute' instead.</value>
   </data>
+  <data name="PreferNotUsingAlwaysTrueAssertConditionsAnalyzerTitle" xml:space="preserve">
+    <value>Always-passing assert</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: Assertion condition is always true

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 7]
```diff
@@ -454,4 +454,10 @@
...
+  </data>
+  <data name="PreferNotUsingAlwaysTrueAssertConditionsAnalyzerMessageFormat" xml:space="preserve">
+    <value>Always-passing 'Assert.{0}' assert</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: Review or remove the assertion as its condition is known to be always true.

**Review by Evangelink — APPROVED** [review]
A couple of renames but otherwise the logic and tests are good! Thanks @engyebrahim

**Review by Evangelink — APPROVED** [review]

**Pull request merged**